qiime2/sdk/parsl_config.py

```diff
@@ -0,0 +1,193 @@
+            # appdirs.user_config_dir(appname='qiime2', author='...')
+            if os.path.exists(fp_ := os.path.join(
+                    appdirs.user_config_dir('qiime2'), 'qiime2_config.toml')):
+                config_fp = fp_
+            # 3. Check in admin writable location
+            # /etc/
+            # site_config_dir
+            # appdirs.site_config_dir(appname='qiime2, author='...')
+            elif config_fp is None and os.path.exists(fp_ := os.path.join(
+                    appdirs.site_config_dir('qiime2'), 'qiime2_config.toml')):
+                config_fp = fp_
+            # 4. Check in conda env
+            # ~/miniconda3/env/{env_name}/conf
+            elif os.path.exists(fp_ := VENDORED_FP):
+                config_fp = fp_
+            # 5. Write the vendored config to the vendored location and use
+            # that
+            else:
+                with open(VENDORED_FP, 'w') as fh:
+                    tomlkit.dump(VENDORED_CONFIG, fh)
+                config_fp = VENDORED_FP
+
+    config_dict = _get_config(config_fp)
+    mapping = _get_mapping(config_dict)
+
+    processed_config = _process_config(config_dict)
+    config = Config(**processed_config)
+
+    PARSL_CONFIG.parsl_config = config
+    PARSL_CONFIG.action_executor_mapping = mapping
+
+
+def _get_config(fp):
+    """Takes a config filepath and determines if the file exists and if so if
+    it contains parsl config info.
+    """
+    with open(fp, 'r') as fh:
+        config_dict = tomlkit.load(fh)
+
+    return config_dict.get('parsl')
+
+
+def _get_mapping(config_dict):
+    """Takes a config filepath and determines if the file exists and if so if
+    it contains action executor mapping info.
+    """
+    return config_dict.pop('executor_mapping', {})
+
+
+def _process_config(config_dict):
+    """Takes a path to a toml file describing a parsl.Config object and parses
+    it into a dictionary of kwargs that can be used to instantiate a
+    parsl.Config object.
+    """
+    config_kwargs = {}
+
+    for key, value in config_dict.items():
+        # Our value is a None
+        if isinstance(value, str) and value.lower() == 'none':
+            config_kwargs[key] = None
+        # We have a list of values
+        elif isinstance(value, list):
+            config_kwargs[key] = []
+            for item in value:
+                config_kwargs[key].append(_process_key(key, item))
+        # We have a single value
+        else:
+            config_kwargs[key] = _process_key(key, config_dict[key])
+
+    return config_kwargs
+
+
+def _process_key(key, value):
+    """Takes a key given in the parsl config file and turns its value into the
+    correct data type or class instance to be used in instantiating a
+    parsl.Config object.
+    """
+    # Our key needs to point to some object.
+    if key in module_paths:
+        module = importlib.import_module(module_paths[key])
+        class_ = getattr(module, value.pop('class'))
```

Comment:
not important... like at all. But people generally get around this by using either `cls` or `klass` depending on context/tradition.

Reply:
(the toml key itself is obviously fine)